k2/csrc/array.h

```diff
@@ -52,70 +56,150 @@ struct Array2Size {
-             // object.
-
-  Array2<I, Ptr> operator[](I i) {
-    // TODO(haowen): fill real data here
-    Array2<I, Ptr> array;
+             // supporting operator [].  data[indexes2[indexes1[0]]] through
+             // data[indexes2[indexes1[size] - 1]] must be accessible through
+             // this object.
+
+  Array2<Ptr, I> operator[](I i) const {
+    DCHECK_GE(i, 0);
+    DCHECK_LT(i, size);
+
+    Array2<Ptr, I> array;
+    array.size = indexes1[i + 1] - indexes1[i];
+    array.indexes = indexes2 + indexes1[i];
+    array.data = data;
     return array;
   }
 };
 
 // Note: we can create Array4 later if we need it.
 
+/*
+  `DataPtrCreator::Create` wraps data storage std::unique_ptr<value_type(Ptr)>
+  as a `Ptr` and returns it, this class is just for test purpose for now.
+*/
+template <typename ValueType, typename I>
+struct DataPtrCreator;
+
+// General case where stride is 1 and the underlying data manager is a raw
+// pointer, we will just return data_storage.get()
+template <typename ValueType, typename I>
+struct DataPtrCreator<ValueType *, I> {
+  static ValueType *Create(const std::unique_ptr<ValueType> &data_storage,
+                           int32_t stride) {
+    CHECK_EQ(stride, 1);
+    return data_storage.get();
+  }
+};
+
+// Specialized case where stride is bigger than 1 and the underlying data
+// manager is StridedPtr, we will create a StridedPtr with data_storage and
+// return it.
+template <typename ValueType, typename I>
+struct DataPtrCreator<StridedPtr<ValueType, I>, I> {
+  static StridedPtr<ValueType, I> Create(
+      const std::unique_ptr<ValueType> &data_storage, int32_t stride) {
+    CHECK_GT(stride, 1);
+    StridedPtr<ValueType, I> strided_ptr(data_storage.get(), stride);
+    return strided_ptr;
+  }
+};
+
+// Allocate memory for Array2, this class is just for test purpose.
+// It will be usually called after those `GetSizes` functions (e.g.
+// FstInverter.GetSizes) to allocate memory in test code.
+template <typename Ptr, typename I>
+struct Array2Storage {
+  using ValueType = typename Array2<Ptr, I>::ValueType;
+  explicit Array2Storage(const Array2Size<I> &array2_size, int32_t stride)
+      : indexes_storage_(new I[array2_size.size1 + 1]),
+        data_storage_(new ValueType[array2_size.size2 * stride]) {
+    array2_size_ = array2_size;
+    array2_.size = array2_size_.size1;
+    array2_.indexes = indexes_storage_.get();
+    array2_.data = DataPtrCreator<Ptr, I>::Create(data_storage_, stride);
+  }
+
+  void FillIndexes(const std::vector<I> &indexes) {
+    CHECK_EQ(indexes.size(), array2_size_.size1 + 1);
+    for (auto i = 0; i != indexes.size(); ++i) array2_.indexes[i] = indexes[i];
+  }
+
+  void FillData(const std::vector<ValueType> &data) {
+    CHECK_EQ(data.size(), array2_size_.size2);
+    for (auto i = 0; i != array2_size_.size2; ++i) array2_.data[i] = data[i];
+  }
+
+  Array2<Ptr, I> &GetArray2() { return array2_; }
+
+ private:
+  Array2<Ptr, I> array2_;
```

Comment:
I'd just call these array_ and size_ ?

Reply:
Sure (removed size_ as we now have `size2` in Array2)